HIP/hip-868.md

```diff
@@ -0,0 +1,155 @@
+the ether balance would be sent to the beneficiary. This moved the operation
+into a "sweep" operation that allowed contracts to clean up all balance without
+having to query it. For contracts that did not exist prior to the transaction
+SELFDESTRUCT would continue to operate normally. Balance would be sent to the
+beneficiary, and code and storage would be deleted. Since none of this data had
+been part of the committed data the deletion did not impact any plans for
+changes to their storage model.
+
+Within the EVM Hedera should support a similar model, as contracts will now
+expect SELFDESTRUCT to operate as a sweep in some cases. Outside the EVM,
+however, there is no need to change any semantics. HAPI calls can still be used
+to delete a contract and state expiry can still delete an account.
+
+## Rationale
+
+Hedera will support the [EIP-6780](https://eips.ethereum.org/EIPS/eip-6780)
+change to the SELFDESTRUCT operation. If the contract existed prior to the
+transaction the contract will not be destroyed and the contents will instead be
+swept to the beneficiary. Contracts within the same transaction will be
+destroyed as per prior behavior.
+
+However, since Hedera places "ownership" of token balances with the account
+holding the balance we will need to extend the behavior of the sweep mode to
+also transfer all HTS tokens to the beneficiary. Notably, for a successful
+transaction beneficiary accounts must support the receiving of HTS tokens via
+association configuration. This is consistent with current self-destruct
+behavior and consistent with the spirit of the Ethereum change in that code and
+the account are no longer deleted.
+
+When the self-destruct is invoked on a contract within a transaction the
+behavior of Hedera will remain as it currently is in all respects.
+
+## User stories
+
+* As a smart contract developer I want to use the "sweep" semantics of EIP-6780
+  in my contracts to sweep out value (HBAR and HTS token balance) while not
+  destroying the contract.
+* As a smart contract developer I want to be able to continue to use the
+  transient contract pattern and have it behave the same way prior to this HIP.
+* As a HAPI user I want to be able to control my contract account in the same
+  ways I did prior to this HIP.
+
+## Specification
+
+There are no changes needed to the JSON-RPC relay.
+
+### EVM Implementation
+
+The SelfDestructOperation will be re-implemented with hedera semantics. An
+existing hedera version of self-destruct exists that may be manipulated in a way
+similar to how the Besu EVM's implementation was altered.
+
+The existing transient self-destruct behaviors will be retained, and the new
+sweep mode changes will be added. All hBar and HTS tokens will be swept as part
+of the self-destruct to the beneficiary address, what will change is that the
+contract will not be marked for removal.
+
+### Hedera Services
+
+Hedera services may need to be changed to ensure that the transaction processing
```

Comment:
Q: Does the EVM assign an address to the contract that did not exist prior to transaction processing or does it go unaddressed?

Reply:
Another question is should that contract call another contract or send cryptocurrency.
What address does the EVM/Ethereum use to express the from address in that account to account interaction?
This may inform how we identify the transient contract

Reply:
You know the address of the contract during the initcode, so yes.  The case it matters for 6780 would be CREATE2 initcode, where the address is a function within the EVM of the salt and initcode.  We are safe not allocating a hedera address until after the full transaction completes (until the hedera account service precompile launches...)

The other case is contract creation transaction.  That already has a high overhead cost to users.

Reply:
The problem is if we don't create an account then the record file will have an ambiguous contractId as the recipient of any transfers as part of the sweep.

If we can get a unique address that will never be reused then we should issue an contractId with the record files showing ContractCreate, Contract Delete child transactions